Reduce Amount of Steps in the MOX

dmallory42 opened the pull request:
Fixes #8311

#### Changes proposed in this Pull Request

Note: This PR relates to the changes to the MOX identified in https://github.com/Automattic/woocommerce-payments/issues/8311. The scope of this PR is everything that happens **after** the Connect page/Jetpack setup, and **before** the redirection to the Stripe KYC. Connect page changes are addressed in #8378, and Jetpack changes will be addressed in a future piece of work.

* Reduces the amount of steps in the Onboarding MOX, removing the personal details and mode selection. (Noting that mode selection now happens earlier, implemented in #8378.)
* Styling updates to the form design.
* Adding the TOS note to the MOX. 

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Screenshots

To follow

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Spin up a new JN site from this branch using the `botwoo` below.
* Go to `Payments` and click `Finish Setup`.
* Connect Jetpack.
* Verify that the new onboarding flow looks like the designs (https://github.com/Automattic/woocommerce-payments/issues/8311).
* Test the form (e.g. trying to submit with inputs unselected, selecting then reselecting values and making sure the behaviour is as expected/consistent). 
* Continue to the Stripe KYC. Create an account and verify that it is successful, and that the values selected in the WooPayments MOX e.g. Country are either prefilled or not asked for again in the Stripe KYC.
* Test the flow again using a new JN site using mobile emulation in the browser and verify it works as expected.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

dmallory42 committed:
Update changelog.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8426` or branch name `dev/8311-reduce-mox` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=dev/8311-reduce-mox) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: e5660ec9b5c4b3e24ee5d6e367ebf600da9cde0b
- Build time: 2024-03-22 17:40:26 UTC

_Note: the build is updated when a new commit is pushed to this PR._

dmallory42 committed:
Styling changes to match designs.

dmallory42 committed:
Fixing style on button to match designs.

dmallory42 committed:
Updates to the styling and functionality of the MOX.

dmallory42 committed:
Merge branch 'develop' into dev/8311-reduce-mox

dmallory42 committed:
Merge branch 'develop' into dev/8311-reduce-mox

oaratovskyi commented:
@dmallory42 thanks for working on this!
So far I tested the PR and found next things in design:
- Change button should be positioned right according to the design
<img width="430" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/79862886/af5b4125-e7b9-44b2-a6b7-0a8969dbe6d6">
<img width="430" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/79862886/ff03fd94-3e10-4acd-a59d-9ccbc7133f30">

- The text should be `Select an option` rather than `Select the legal structure...`
- Color of `What type of goods or services does your business sell?` and the selected MCC should be black, not grey
<img width="436" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/79862886/921b3c31-4982-4533-ab6f-a68fe05db6b0">
<img width="430" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/79862886/3f38ee8c-8502-441a-ae95-10bc92d073a7">

-  nitpick: in including `WooPay merchant terms` WooPay also should be a part of the link

Regarding the code - the prefill is broken both with PO and non-PO cases

oaratovskyi commented on client/onboarding/index.tsx:
```diff
@@ -66,6 +56,7 @@ const OnboardingStepper = () => {
 	);
 };
 
+// TODO: WE CAN PROBABLY JUST ATTACH THIS ON A SERVER LEVEL (URL AND BUSINESS DATA)
```
hmm, interesting idea, we may
but it's okay this way I guess

dmallory42 replied:
Sorry - this is just a TODO note for myself about how we could consider prepopulating the URL and Business name (which we used to ask for in the MOX) when making the server request, because we have this info already. I will remove it or make a new issue for it before merging. 👍

oaratovskyi replied:
no worries, gotcha!

oaratovskyi commented on client/onboarding/step.tsx:
```diff
@@ -32,7 +31,6 @@ const Step: React.FC< Props > = ( { name, children } ) => {
 
 	return (
 		<>
-			<div className="stepper__progress" style={ { width } } />
```
oh no :( I liked progress

dmallory42 replied:
It wasn't in the design, which is why I removed it - I guess it probably isn't as necessary now that we have only 2 steps.

cc. @elizaan36 just to double check this!

elizaan36 replied:
So... The reason I think the progress bar doesn't work as well for these two screens is it's somewhat misleading. I think the progress bar would only really work if it was continuous through a contained flow, from 1. WordPress.com connection, 2. MOX, 3. Stripe verification. I'm open to other ideas though, if you think it's essential for this section of the onboarding!

dmallory42 replied:
Thanks for sharing the thinking behind it. I agree it can be misleading because of the extra steps like Jetpack Connection and Stripe KYC. Will have a think about if there's another way to show progress, but given the MOX is now only 2 steps it doesn't seem necessary imo :)

oaratovskyi replied:
Thanks! Closing this one

oaratovskyi commented on client/onboarding/steps/business-details.tsx:
```diff
@@ -9,22 +9,21 @@ import React from 'react';
 	getMccsFlatList,
 } from 'onboarding/utils';
 import { BusinessType } from 'onboarding/types';
+import InlineNotice from 'wcpay/components/inline-notice';
```
```suggestion
import InlineNotice from 'components/inline-notice';
```

oaratovskyi commented on client/onboarding/steps/business-details.tsx:
```diff
@@ -9,22 +9,21 @@ import React from 'react';
 } from 'onboarding/utils';
 import { BusinessType } from 'onboarding/types';
+import InlineNotice from 'wcpay/components/inline-notice';
+import strings from 'wcpay/onboarding/strings';
```
```suggestion
import strings from 'onboarding/strings';
```

oaratovskyi commented on client/onboarding/style.scss:
```diff
@@ -65,6 +55,7 @@ body.wcpay-onboarding__body {
 
 			&__heading {
 				@include wp-title-large;
+				font-family: 'SF Pro Display', $default-font;
```
I am not sure if we should change font-family, I didn't find any occurrences in the codebase like this

dmallory42 replied:
The context for this is in this P2 post (paJDYF-ci2-p2), specifically:

> The designs are using SF Pro Display at 32px, which is a slight change from the existing SF Pro Text used for the headers. The nuance in letterspacing makes the headlines more readable for larger sizes.

I think it should hopefully be safe to do so, because as far as I can tell the only difference with the "Display" font is that it renders more smoothly for large font sizes. It'll just fall back to the default font set (which is SF Pro on OSX devices, and then falling back to equivalent fonts on other devices) if there's any problem.

oaratovskyi replied:
thanks for explaining! I'm good with that!

oaratovskyi reviewed (commented):
code comments so far
will take a look one more time when tests are done!

dmallory42 commented:
Thanks a lot for looking, @oaratovskyi! 

> Change button should be positioned right according to the design

Yep, this one is on my list to look at, thanks for mentioning it 🙂 

> The text should be `Select an option` rather than `Select the legal structure...`

I wanted to double check with @elizaan36 whether this change on the design is an intentional one. Currently, we have some slightly more specific placeholder messages like "Select the legal structure of your business" (see screenshot) but in the design we have the more generic "Select an option" - is that intended? 🙂 

<img width="541" alt="Screenshot 2024-03-21 at 15 41 33" src="https://github.com/Automattic/woocommerce-payments/assets/11770181/88f5275c-0671-421c-9ab9-523c3870648b">

> Color of What type of goods or services does your business sell? and the selected MCC should be black, not grey

> nitpick: in including WooPay merchant terms WooPay also should be a part of the link

Good catches, thanks!

Will take a look at the things you mentioned before requesting full review. Thanks!

dmallory42 committed:
Update client/onboarding/steps/business-details.tsx

Good catch!

Co-authored-by: Oleksandr Aratovskyi <79862886+oaratovskyi@users.noreply.github.com>

dmallory42 committed:
Update client/onboarding/steps/business-details.tsx

Co-authored-by: Oleksandr Aratovskyi <79862886+oaratovskyi@users.noreply.github.com>

dmallory42 committed:
Updates to unit tests.

dmallory42 committed:
Merge branch 'dev/8311-reduce-mox' of github.com:Automattic/woocommerce-payments into dev/8311-reduce-mox

dmallory42 committed:
Merge branch 'develop' into dev/8311-reduce-mox

dmallory42 committed:
Fix tests.

dmallory42 committed:
Styling fixes.

dmallory42 commented:
> Regarding the code - the prefill is broken both with PO and non-PO cases

Can you check again? It works for me in both cases 🤔 

Remember we're now prefilling less data - so for example Email and Name won't be prefilled, but country / MCC / business type should all still be prefilled.

dmallory42 commented:
I've addressed all the comments apart from the positioning of the `Change` button (will ask about it in Slack) and updated the tests, so I'm going to open this up for review!

elizaan36 commented:
> Currently, we have some slightly more specific placeholder messages like "Select the legal structure of your business" (see screenshot) but in the design we have the more generic "Select an option" - is that intended?

Yes that was intentional. The helper text is redundant with the form label and feels more tedious to read. I think the form label communicates everything the user needs to know and the more generic "Select an option" simplifies things. WDYT?

elizaan36 commented:
Here are some notes and questions. 

1. This notice was an interesting one. I closed out of the MOX and then when I tried to restart this notice appeared. It persisted when I continued through to the next screen, which was a little strange. I don't think it's needed as it doesn't really help the user anymore. Can we remove it or am I missing something?
<img width="2948" alt="Frame 1000005933" src="https://github.com/Automattic/woocommerce-payments/assets/5121465/65390013-d774-4e92-b82b-37ec5234fff9">

2. It will make a big difference if we give the form some breathing room from the top of the page. The padding between logo and headline should be bumped up.
<img width="625" alt="Frame 1000005928" src="https://github.com/Automattic/woocommerce-payments/assets/5121465/f57a8737-0183-42ac-aa92-73c19104745a">

3. There's some extra padding between the TOS copy and CTA. The gap should be consistent throughout all of the form elements.
<img width="687" alt="Frame 1000005929" src="https://github.com/Automattic/woocommerce-payments/assets/5121465/3a48af4e-141b-4251-b836-5060b4cdc43a">

4. +1 on @oaratovskyi 's [comment here](https://github.com/Automattic/woocommerce-payments/pull/8426#issuecomment-2012627554) that the change link should be aligned to the right of the saved country box.

5. Dropdowns need some visual tweaks, with the MCC dropdown diverging quite a bit. I shared details in the notes and added the designs to the left.
<img width="1505" alt="Frame 1000005931" src="https://github.com/Automattic/woocommerce-payments/assets/5121465/7eb72e78-6705-47b5-bd7f-4e7cd9129314">
<img width="1894" alt="Frame 1000005932" src="https://github.com/Automattic/woocommerce-payments/assets/5121465/094b87d2-848f-46a1-9f0e-42f18a49d523">

dmallory42 committed:
Merge branch 'develop' into dev/8311-reduce-mox

dmallory42 commented:
> Yes that was intentional. The helper text is redundant with the form label and feels more tedious to read. I think the form label communicates everything the user needs to know and the more generic "Select an option" simplifies things. WDYT?

I agree, makes sense! I wanted to double check before making the change. 🙂 

Thanks for the notes, they are very useful! I'll try and resolve these things today, and if you have time to check again later I'd appreciate it, since if we are able to get the PR ready today we can fit it into the next WooPayments release.

> This notice was an interesting one. I closed out of the MOX and then when I tried to restart this notice appeared. It persisted when I continued through to the next screen, which was a little strange. I don't think it's needed as it doesn't really help the user anymore. Can we remove it or am I missing something?

We can probably remove it now. Originally, this was intended to save the merchant's progress so they could come back to the onboarding later - however, now that it's only two steps, and mostly just selecting from dropdowns, I think it's probably easier to remove it. I'll take care of that! 

Will let you know once I fix the other comments. Thanks again!

dmallory42 committed:
Remove Flow state persistence functionality.

dmallory42 committed:
Merge branch 'dev/8311-reduce-mox' of github.com:Automattic/woocommerce-payments into dev/8311-reduce-mox

oaratovskyi commented on client/onboarding/utils.ts:
```diff
@@ -86,14 +86,3 @@ export const getMccsFlatList = (): ListItem[] => {
 	}, [] as ListItem[] );
 };
-
-export const persistFlowState = (
```
we can also remove these
<img width="643" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/79862886/f103dbeb-544c-4153-9f25-1bfa8f570a68">

oaratovskyi replied:
@dmallory42 import of apiFetch is also can be safely removed 🙂

dmallory42 replied:
Done, thanks!

oaratovskyi commented on client/globals.d.ts:
```diff
@@ -102,10 +102,6 @@ declare global {
 			business_types: Country[];
 			mccs_display_tree: MccsDisplayTreeItem[];
 		};
-		onboardingFlowState?: {
```
please remove `OnboardingFields` from import

oaratovskyi commented on client/onboarding/index.tsx:
```diff
@@ -11,18 +11,13 @@ import { OnboardingContextProvider, useOnboardingContext } from './context';
-import { persistFlowState } from './utils';
 
 const OnboardingStepper = () => {
-	const { data } = useOnboardingContext();
```
please remove `useOnboardingContext` from import

dmallory42 committed:
Adjust padding.

dmallory42 committed:
Making the margin consistent between form elements.

oaratovskyi reviewed (approved):
Tested one more time! Prefill works for those data user specifies
Of course it won't prefill the name surname etc, my bad 🙈 

I think it's almost ready, need final polishing touches design wise, but I pre-approve it!

dmallory42 committed:
Update Select to have more generic messages.

oaratovskyi committed:
Fix CSS bug with business type

oaratovskyi committed:
Merge remote-tracking branch 'origin/dev/8311-reduce-mox' into dev/8311-reduce-mox

dmallory42 committed:
CSS changes to match designs.

dmallory42 committed:
Merge branch 'dev/8311-reduce-mox' of github.com:Automattic/woocommerce-payments into dev/8311-reduce-mox

dmallory42 committed:
Tidying up.

dmallory42 committed:
Merge branch 'develop' into dev/8311-reduce-mox

dmallory42 committed:
Temporarily disable test.

dmallory42 committed:
Merge branch 'dev/8311-reduce-mox' of github.com:Automattic/woocommerce-payments into dev/8311-reduce-mox

dmallory42 committed:
Merge branch 'develop' into dev/8311-reduce-mox

dmallory42 commented:
Thanks @oaratovskyi! I've addressed the comments you left.

This should be ready for design review, @elizaan36 if you could take another look I'd appreciate it! 

The only thing outstanding from the code POV is that I have one test I need to fix (the code works, only the test is broken because of changing option names) - I'll take a short break and then have another look at that.

dmallory42 committed:
Remove import of API fetch.

dmallory42 committed:
Merge branch 'dev/8311-reduce-mox' of github.com:Automattic/woocommerce-payments into dev/8311-reduce-mox

elizaan36 reviewed (approved):
Thanks for making those adjustments, @dmallory42. We still need another ~10px of breathing room at the top. The form feels a bit claustrophobic still, and adding that space will balance it vertically on the page a bit better.

<img width="1625" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/5121465/aa937624-04c6-42d0-acfb-3e386e3ad8a9">

One other thing I wanted to mention. It doesn't happen every time, but clicking the "Back" button will sometimes direct me to the WordPress dashboard rather than the Payments connect page. I'm not sure why, and if it's a result of the testing environment but wanted to mention in case it's something to look into.

Otherwise looking good!

dmallory42 commented:
Thanks @elizaan36! 

> Thanks for making those adjustments, @dmallory42. We still need another ~10px of breathing room at the top. The form feels a bit claustrophobic still, and adding that space will balance it vertically on the page a bit better.

Appreciate the callout on this, I tried to adjust it but because of the way the components sit on the page it was difficult to enter a precise input on the breathing room so it was a bit of a best estimate. I added another ~10px, this is what it looks like now - hope this is okay!:

<img width="632" alt="Screenshot 2024-03-22 at 14 45 22" src="https://github.com/Automattic/woocommerce-payments/assets/11770181/017bb83b-061e-43fe-8247-afe337c912e8"> 

> One other thing I wanted to mention. It doesn't happen every time, but clicking the "Back" button will sometimes direct me to the WordPress dashboard rather than the Payments connect page. I'm not sure why, and if it's a result of the testing environment but wanted to mention in case it's something to look into.

I think maybe this could depend on the entry route into the Onboarding? That's the only thing I can think that would cause it. Would the desired behaviour here to be to always go back to the Payments Connect page? I can open an issue up for it!

dmallory42 committed:
Merge branch 'develop' into dev/8311-reduce-mox

vladolaru reviewed (approved):
I've tested the various fields in the MOX and the business details ones all follow the right logic around stuff like available options tied to the country, etc.

Thanks for working on it, Daniel!

dmallory42 committed:
Merge branch 'dev/8311-reduce-mox' of github.com:Automattic/woocommerce-payments into dev/8311-reduce-mox

dmallory42 committed:
Minor change to the padding.

dmallory42 committed:
Merge remote-tracking branch 'origin/develop' into dev/8311-reduce-mox

elizaan36 commented:
Nice, the height looks good. 🎉

> Would the desired behaviour here to be to always go back to the Payments Connect page?

I think it should always go back to the previous step, which I believe would always be the Payments connect page now that we're making these adjustments to all of the entry points.

oaratovskyi committed:
fix border and background in sandbox panel

dmallory42 committed:
Removing unneeded strings.

dmallory42 commented:
>  I think it should always go back to the previous step, which I believe would always be the Payments connect page now that we're making these adjustments to all of the entry points.

Ah, great - so that _should_ be the current behaviour - so maybe it was just some weirdness with the testing environment. I opened an issue to double check that logic outside of this one!

dmallory42 committed:
Merge branch 'develop' into dev/8311-reduce-mox

dmallory42 committed:
Fix for test.

dmallory42 committed:
Merge branch 'dev/8311-reduce-mox' of github.com:Automattic/woocommerce-payments into dev/8311-reduce-mox
